### PR TITLE
Unify Tophat and Playground servers

### PR DESCRIPTION
**Pull request by BPScott**
### WHY are these changes introduced?

The tophat server is kinda clunky. The most annoying of its offences is that you need to restart the server every time you make a change to a markdown file which takes aaages. It's also separate to the playground server which means people don't always remember it exists.

### WHAT is this pull request doing?

Merges the Tophat server functionality into the playground server.

By loading example markdown files through a webpack loader instead of loading them as serialised data in the html page we can serve a static page as the content and include the markdown as part of the webpack bundle.

This means that you no longer have to restart the server every time you modify an example - hot reloading works for example pages if you modify a component or the example in a README (way better than having to restart the whole server!)

### How to 🎩

- Ensure  the `yarn dev` playground continues to work as expected. Hot reloading when you modify TS file should still work
- Go to /examples within the playground server and see it renders the example list from the markdown
  - Modify the contents of the component, refresh the page in your browser and see that the change is shown
  - Modify an example in that component's README.md, refresh the page in your browser and see that the change is shown.
- Ensure percy and shrink-ray changes look good on this PR. Shrink-ray is expected to grow due to the example code now being compiled into main.js
- Ensure the deployed examples still work on Heroku

At no point should you need to kill and restart your `yarn dev` server to see changes.

### CODE REVIEW

**Comment by kaelig** [comment]
I love it! @amrocha can we reuse any of this in the style guide?

**Comment by amrocha** [comment]
I talked to Ben about this earlier. Yeah, we can use the same technique on the styleguide to process MD files through a webpack loader instead and get hot reloading that way. Once this is up and running with no problems for polaris-react we'll look into getting it into the styleguide as well

**Comment by BPScott** [comment]
It's worth pointing out that the current loader implementation is kinda hacky - it loads all files in a single call when what we really want most of the time is to load a single file's contents. This was done as I didn't want to rewrite the parseMarkdown logic  just yet.

**Comment by BPScott** [comment]
Pushed some new work that unifies the playground and tophat servers (which gives us hot reloading along the way). `yarn tophat` is dead, long live `yarn dev`.

**Comment by amrocha** [comment]
Looks like Heroku is good too :shipit:

File: playground/webpack.config.js [hunk 1]
```diff
@@ -4,23 +4,40 @@ const {
...
-  mode: 'development',
-  devtool: 'eval',
-  stats: {warnings: false},
+  mode: env.production ? 'production' : 'development',
+  devtool: env.production ? 'source-map' : 'eval',
+  stats: {
+    // When transpiling TS using isolatedModules, the compiler doesn't strip
+    // out exported types as it doesn't know if an item is a type or not.
+    // Ignore those warnings as we don't care about them.
+    warningsFilter: /export .* was not found in/,
```
[the remark this answers is not in the document]
Comment (BPScott) [reply playground/webpack.config.js]: Annoyingly this has to go in the file twice  - once for compiles (at the top level) and once for the dev server output (in the devServer block). Oh well

File: playground/webpack.config.js [hunk 2]
```diff
@@ -4,23 +4,40 @@ const {
...
 const ICON_PATH_REGEX = /icons\//;
 const IMAGE_PATH_REGEX = /\.(jpe?g|png|gif|svg)$/;
 
-module.exports = {
+module.exports = (env = {production: false}) => ({
   target: 'web',
-  mode: 'development',
-  devtool: 'eval',
-  stats: {warnings: false},
+  mode: env.production ? 'production' : 'development',
```
[the remark this answers is not in the document]
Comment (BPScott) [reply playground/webpack.config.js]: Yep, they were basically the same beforehand. This unifies them into one - with some logic switching based upon the env you pass in - so we don't need to make modifications in multiple places.

Really happy they I've managed to go from 3 webpack config files (tophat dev, tophat "build a shrink-ray report" version and playground dev) into a single config.

File: playground/Playground.tsx [hunk 3]
```diff
@@ -3,10 +3,15 @@ import {Page} from '@shopify/polaris';
 
 interface State {}
 
-export default class Playground extends React.Component<never, State> {
+export default class Playground extends React.Component<{}, State> {
```
[the remark this answers is not in the document]
Comment (BPScott) [reply playground/Playground.tsx]: Apparently so. It was complaining prior to this as the react router passes props into components.

**Pull request merged**